br/pkg/checkpoint/storage.go

```diff
@@ -0,0 +1,333 @@
+		endIdx := startIdx + CheckpointIdMapBlockSize
+		if endIdx > len(data) {
+			endIdx = len(data)
+		}
+		if err := fn(segmentId, data[startIdx:endIdx]); err != nil {
+			return errors.Trace(err)
+		}
+		startIdx = endIdx
+	}
+	return nil
+}
+
+func chunkInsertCheckpointSQLs(dbName, tableName string, data []byte) ([]string, [][]any) {
+	sqls := make([]string, 0, len(data)/CheckpointIdMapBlockSize+1)
+	argss := make([][]any, 0, len(data)/CheckpointIdMapBlockSize+1)
+	uuid := uuid.New()
+	_ = chunkInsertCheckpointData(data, func(segmentId uint64, chunk []byte) error {
+		sqls = append(sqls, fmt.Sprintf(insertCheckpointSQLTemplate, dbName, tableName))
+		argss = append(argss, []any{uuid[:], segmentId, chunk})
+		return nil
+	})
+	return sqls, argss
+}
+
+type tableCheckpointStorage struct {
+	se               glue.Session
+	checkpointDBName string
+}
+
+func (s *tableCheckpointStorage) flushCheckpointData(ctx context.Context, _ string, data []byte) error {
+	sqls, argss := chunkInsertCheckpointSQLs(s.checkpointDBName, checkpointDataTableName, data)
+	for i, sql := range sqls {
+		args := argss[i]
+		if err := s.se.ExecuteInternal(ctx, sql, args...); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (s *tableCheckpointStorage) flushCheckpointChecksum(ctx context.Context, _ string, data []byte) error {
+	sqls, argss := chunkInsertCheckpointSQLs(s.checkpointDBName, checkpointChecksumTableName, data)
+	for i, sql := range sqls {
+		args := argss[i]
+		if err := s.se.ExecuteInternal(ctx, sql, args...); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func mergeSelectCheckpoint(
+	ctx context.Context,
+	execCtx sqlexec.RestrictedSQLExecutor,
+	dbName, tableName string,
+) ([][]byte, error) {
+	rows, _, errSQL := execCtx.ExecRestrictedSQL(
+		kv.WithInternalSourceType(ctx, kv.InternalTxnBR),
+		nil,
+		selectCheckpointSQLTemplate,
+		dbName, tableName,
+	)
+	if errSQL != nil {
+		return nil, errors.Annotatef(errSQL, "failed to get checkpoint data from table %s.%s", dbName, tableName)
+	}
+
+	var (
+		retData         [][]byte = make([][]byte, 0, len(rows))
+		rowData                  = []byte{}
+		lastUUID        []byte   = nil
+		lastUUIDInvalid bool     = false
+		nextSegmentID   uint64   = 0
+	)
+	for _, row := range rows {
+		uuid, segment_id, data := row.GetBytes(0), row.GetUint64(1), row.GetBytes(2)
+		if len(uuid) == 0 {
+			log.Warn("get the empty uuid, but just skip it")
+			continue
+		}
+		if !bytes.Equal(uuid, lastUUID) {
+			if !lastUUIDInvalid && len(rowData) > 0 {
+				retData = append(retData, rowData)
+			}
+			rowData = make([]byte, 0)
+			lastUUIDInvalid = false
+			nextSegmentID = 0
+			lastUUID = uuid
+		}
+
+		if lastUUIDInvalid {
+			continue
+		}
+
+		if nextSegmentID != segment_id {
+			lastUUIDInvalid = true
+			continue
+		}
+
+		rowData = append(rowData, data...)
+		nextSegmentID += 1
+	}
+	if !lastUUIDInvalid && len(rowData) > 0 {
+		retData = append(retData, rowData)
```

Comment:
Should we return an error if `lastUUIDInvalid`? Or would you explain how the UUID will mismatch in the comments?